Source/AtomicEditor/Application/AEEditorCommon.cpp

```diff
@@ -174,46 +177,105 @@ bool AEEditorCommon::CreateDefaultPreferences(String& path, JSONValue& prefs)
-    editorWindow["maximized"] = true;
+    GetDefaultWindowPreferences(editorWindow);
 
     JSONValue playerWindow;
-    playerWindow["x"] = 0;
-    playerWindow["y"] = 0;
-    playerWindow["width"] = 0;
-    playerWindow["height"] = 0;
-    playerWindow["monitor"] = 0;
-    playerWindow["maximized"] = false;
+    GetDefaultWindowPreferences(playerWindow);
 
     root["editorWindow"] = editorWindow;
     root["playerWindow"] = playerWindow;
 
-    SharedPtr<File> file(new File(context_, path, FILE_WRITE));
+    prefs = root;
 
-    if (!file->IsOpen())
+    SavePreferences(prefs);
+
+    return true;
+}
+
+bool AEEditorCommon::ReadPreferences()
+{
+    FileSystem* fileSystem = GetSubsystem<FileSystem>();
+    String path = GetPreferencesPath();
+
+    JSONValue prefs;
+
+    LoadPreferences(prefs);
+
+    if (!prefs.IsObject() || !prefs["editorWindow"].IsObject())
     {
-        LOGERRORF("Unable to open Atomic Editor preferences for writing: %s", path.CString());
-        return false;
+        if (!CreateDefaultPreferences(path, prefs))
+            return false;
     }
 
-    jsonFile->Save(*file, "   ");
+    JSONValue& editorWindow = prefs["editorWindow"];
 
-    prefs = root;
+    engineParameters_["WindowPositionX"] = editorWindow["x"].GetUInt();
+    engineParameters_["WindowPositionY"] = editorWindow["y"].GetUInt();
+    engineParameters_["WindowWidth"] = editorWindow["width"].GetUInt();
+    engineParameters_["WindowHeight"] = editorWindow["height"].GetUInt();
+    engineParameters_["WindowMaximized"] = editorWindow["maximized"].GetBool();
 
     return true;
 }
 
-bool AEEditorCommon::ReadPreferences()
+void AEEditorCommon::ValidateWindow()
+{
+    Graphics* graphics = GetSubsystem<Graphics>();
+    IntVector2 windowPosition = graphics->GetWindowPosition();
+    int monitors = graphics->GetNumMonitors();
+    IntVector2 maxResolution;
+
+    for (int i = 0; i < monitors; i++)
+    {
+        IntVector2 monitorResolution = graphics->GetMonitorResolution(i);
+        maxResolution += monitorResolution;
+    }
+
+    if (windowPosition.x_ >= maxResolution.x_ || windowPosition.y_ >= maxResolution.y_ || windowPosition.x_ < 0 || windowPosition.y_ < 0)
+    {
+        JSONValue prefs;
+
+        if (!LoadPreferences(prefs))
+            return;
+
+        JSONValue window;
+        GetDefaultWindowPreferences(window);
+
+        prefs[context_->GetEditorContext() ? "editorWindow" : "playerWindow"] = window;
+
+        graphics->SetMode(0, 0);
+        graphics->CenterWindow();
+        if (context_->GetEditorContext())
+        {
+            graphics->Maximize();
+        }
+
+        SavePreferences(prefs);
+    }
+}
+
+void AEEditorCommon::GetDefaultWindowPreferences(JSONValue& windowPrefs)
+{
+    windowPrefs["x"] = 0;
+    windowPrefs["y"] = 0;
+    windowPrefs["width"] = 0;
+    windowPrefs["height"] = 0;
+    windowPrefs["monitor"] = 0;
+    windowPrefs["maximized"] = context_->GetEditorContext() ? true : false;
```

Comment:
You cannot get both the editor and player preferences like this, as it will either always be true (editor process) or false (player process)